src/main/java/org/folio/rest/impl/CheckinReceivePiecesHelper.java

```diff
@@ -657,4 +641,48 @@ private boolean isMissingLocation(PoLine poLine, Piece piece) {
 
   abstract String getLocationId(Piece piece);
 
+  /**
+   * Updates items in the inventory storage with check-in/receiving details if any. On
+   * success updates corresponding records as received
+   *
+   * @return {@link CompletableFuture} which holds map with PO line id as key
+   *         and list of corresponding pieces as value
+   */
+  CompletableFuture<Map<String, List<Piece>>> updateInventoryItems(Map<String, Map<String, String>> pieceLocationsGroupedByPoLine, Map<String, List<Piece>> piecesGroupedByPoLine) {
+    // Collect all piece records with non-empty item ids. The result is a map
+    // with item id as a key and piece record as a value
+    Map<String, Piece> piecesWithItems = collectPiecesWithItemId(piecesGroupedByPoLine);
+    List<String> polineIds = new ArrayList<>(piecesGroupedByPoLine.keySet());
+
+    // If there are no pieces with ItemId, continue
+    if (piecesWithItems.isEmpty()) {
+      return completedFuture(piecesGroupedByPoLine);
+    }
+
+    return getItemRecords(piecesWithItems)
+      .thenCombine(getPoLines(polineIds), (items, poLines) -> {
+        List<CompletableFuture<Boolean>> futuresForItemUpdates = new ArrayList<>();
+        for (JsonObject item : items) {
+          String itemId = item.getString(ID);
+          Piece piece = piecesWithItems.get(itemId);
+
+          PoLine poLine = searchPoLineById(poLines, piece.getPoLineId());
+          String pieceLocation = pieceLocationsGroupedByPoLine.get(poLine.getId()).get(piece.getId());
+
+          futuresForItemUpdates.add(updateHoldingAndItem(item, piece, poLine, pieceLocation));
```

Comment:
I've tested this by API tests and there are 2 issues:
1. User would like to receive a few pieces with associated items but changing location for all of them to new one and there is no holdings record available yet for combination of this location + instance. The logic will search for a holding for each item independently. As a result in current scenario more than one holdings record will be created. In my local run I tried to receive 10 pieces changing their location to `758258bc-ecc1-41b8-abca-f7b610822ffd` and 10 holdings records were created.
2. User would like to receive a few pieces with associated items but changing location for all of them to which does not exists in inventory storage e.g. location was deleted right before receiving. At the moment the logic:
   - searches for a holdings record and does not find any
   - tries to create a new one and inventory storage rejects it because there is no such location i.e. error is like following:
      > ErrorMessage(fields=Map(Detail -> Key (permanentlocationid)=(eb2d063a-5b4c-4cab-8db1-5fc5c5941df6) is not present in table "location"., s -> diku_mod_inventory_storage, n -> holdings_record_permanentlocationid_fkey, t -> holdings_record, Line -> 3266, File -> ri_triggers.c, SQLSTATE -> 23503, Routine -> ri_ReportViolation, V -> ERROR, Message -> insert or update on table "holdings_record" violates foreign key constraint "holdings_record_permanentlocationid_fkey", Severity -> ERROR))

   - the `updateHoldingAndItem` method defined above does not add any error but just returns `false` as a result of future
   - the `storeUpdatedPieceRecords` defined in ReceivingHelper checks that there is no any error associated with any piece and sends piece update requests

Reply:
Ignore the previous comment because I did not get the latest changes and used previous version of branch. Now I updated and see a bit different results. Now when I change location, there is no holdings search requests. This is because of update in `holdingUpdateOnCheckinReceiveRequired` method:
```java
&& StringUtils.equals(locationId, piece.getLocationId());
```

Reply:
fixed holdings creation